airflow/plugins/operators/scrape_state_geoportal.py

```diff
@@ -0,0 +1,206 @@
+                    f"Downloaded {self.product} data with {len(all_features)} rows!"
+                )
+
+                return all_features
+
+        except requests.exceptions.RequestException as e:
+            logging.info(f"An error occurred: {e}")
+
+            raise
+
+
+# # Function to convert coordinates to WKT format
+def to_wkt(geometry_type, coordinates):
+    if geometry_type == "LineString":
+        # Format as a LineString
+        coords_str = ", ".join([f"{lng} {lat}" for lng, lat in coordinates])
+        return f"LINESTRING({coords_str})"
+    elif geometry_type == "MultiLineString":
+        # Format as a MultiLineString
+        multiline_coords_str = ", ".join(
+            f"({', '.join([f'{lng} {lat}' for lng, lat in line])})"
+            for line in coordinates
+        )
+        return f"MULTILINESTRING({multiline_coords_str})"
+    else:
+        return None
+
+
+class JSONExtract(StateGeoportalAPIExtract):
+    bucket = API_BUCKET
+
+
+class StateGeoportalAPIOperator(BaseOperator):
+    template_fields = (
+        "root_url",
+        "service",
+        "product",
+        "where",
+        "outFields",
+        "f",
+        "resultRecordCount",
+    )
+
+    def __init__(
+        self,
+        root_url,
+        service,
+        product,
+        where,
+        outFields,
+        f,
+        resultRecordCount,
+        **kwargs,
+    ):
+        self.root_url = root_url
+        self.service = service
+        self.product = product
+        self.where = where
+        self.outFields = outFields
+        self.f = f
+        self.resultRecordCount = resultRecordCount
+
+        """An operator that extracts and saves JSON data from the State Geoportal
+            and saves it as one JSONL file, hive-partitioned by date in Google Cloud
+        """
+
+        # Save JSONL files to the bucket
+        self.extract = JSONExtract(
+            root_url=self.root_url,
+            service=self.service,
+            product=f"{self.product}_geodata",
+            where=self.where,
+            outFields=self.outFields,
+            f=self.f,
+            resultRecordCount=self.resultRecordCount,
+            filename=f"{self.product}_geodata.jsonl.gz",
+        )
+
+        super().__init__(**kwargs)
+
+    def execute(self, **kwargs):
+        api_content = self.extract.fetch_from_state_geoportal()
+
+        df = pd.json_normalize(api_content)
+
+        if self.product == "state_highway_network":
+            # Select and rename columns
+            columns = {
+                "properties.Route": "Route",
+                "properties.County": "County",
+                "properties.District": "District",
+                "properties.RouteType": "RouteType",
+                "properties.Direction": "Direction",
+                "geometry.type": "type",
+                "geometry.coordinates": "coordinates",
+            }
+            df = df[list(columns.keys())].rename(columns=columns)
```

Comment:
**suggestion (non-blocking):** I don't think this is something that needs to be addressed right now, but it seems like we could do this without hard-coding the `product` here. We're really just cleaning up column names -- removing `properties.` and `geometry.` prefixes that come from the GeoJSON.

Reply:
yes very good point, refactored